### PR TITLE
[DocSupport] Ensure synthesized target is correct for enclosed nominal decls inside other extensions. rdar://29857103

### PR DESCRIPTION
**Pull request by nkcsgexi**

### CODE REVIEW

**Comment by nkcsgexi** [comment]
@swift-ci Please smoke test and merge

**Comment by nkcsgexi** [comment]
@swift-ci please smoke test and merge

File: lib/AST/ASTPrinter.cpp [hunk 1]
```diff
@@ -859,11 +859,16 @@ class PrintAST : public ASTVisitor<PrintAST> {
 
       // Get the innermost nominal type context.
       DeclContext *DC;
-      if (isa<NominalTypeDecl>(Current) || isa<ExtensionDecl>(Current))
+      if (isa<NominalTypeDecl>(Current))
         DC = Current->getInnermostDeclContext();
+      else if (isa<ExtensionDecl>(Current))
+        DC = Current->getInnermostDeclContext()->
```
Comment (slavapestov) [inline lib/AST/ASTPrinter.cpp]: Is it correct to remap ExtensionDecl DCs to the DC of the nominal type here?

File: test/SourceKit/DocSupport/Inputs/cake.swift [hunk 2]
```diff
@@ -79,4 +79,7 @@ public protocol P4 {}
 
 extension C1 : P4 {
   public func C1foo() {}
+  public struct C1S1{
```
Comment (slavapestov) [inline test/SourceKit/DocSupport/Inputs/cake.swift]: Can you add tests for generic nested types and extensions of generic types also?

File: lib/AST/ASTPrinter.cpp [hunk 3]
```diff
@@ -859,11 +859,16 @@ class PrintAST : public ASTVisitor<PrintAST> {
...
+          getAsNominalTypeOrNominalTypeExtensionContext();
       else
         DC = Current->getDeclContext();
 
+      assert(DC->isTypeContext());
```
Comment (slavapestov) [inline lib/AST/ASTPrinter.cpp]: Note that isTypeContext() is true for extension contexts, too.

So it looks like the original assert was this:

```
Assertion failed: (ownerNominal == baseTy->getAnyNominal()), function getContextSubstitutions, file /Volumes/ssd/SSDDocument/stash_master_sourceroot/swift/lib/AST/Type.cpp, line 3139.
```

What was ownerNominal and baseTy in that case? Is there a bug in getContextSubstitutions() when the DC is an extension of the base type? That should work...

Comment (slavapestov) [inline lib/AST/ASTPrinter.cpp]: Basically what I don't understand is how getContextSubstitutions(ExtensionDecl) is different from getContextSubstitutions(ExtensionDecl->getAsNominalTypeOrNominalTypeExtension())

Comment (slavapestov) [inline lib/AST/ASTPrinter.cpp]: Here is the code in getContextSubstitutions():

```
  auto *ownerNominal = dc->getAsNominalTypeOrNominalTypeExtensionContext();
  if (auto *ownerClass = dyn_cast<ClassDecl>(ownerNominal))
    baseTy = baseTy->getSuperclassForDecl(ownerClass, resolver);

  assert(ownerNominal == baseTy->getAnyNominal());
```

Note that regardless of whether you pass in the ExtensionDecl as 'dc' or the NominalTypeDecl, ownerNominal is the same value.

File: tools/SourceKit/lib/SwiftLang/SwiftDocSupport.cpp [hunk 4]
```diff
@@ -199,8 +202,12 @@ class AnnotatingPrinter : public StreamPrinter {
       return;
     unsigned StartOffset = OS.tell();
     initDefaultMapToUse(D);
-    EntitiesStack.emplace_back(D, SynthesizeTarget, getDefaultImplementation(D),
-                               StartOffset, false);
+    const NominalTypeDecl *SynthesizedTarget = nullptr;
+    // If D is declared in the extension, then the synthesized target is valid.
+    if (D->getDeclContext() == SynthesizedExtensionInfo.first)
+      SynthesizedTarget = SynthesizedExtensionInfo.second;
```
Comment (slavapestov) [inline tools/SourceKit/lib/SwiftLang/SwiftDocSupport.cpp]: Is this the real fix? I think you can revert your change in ASTPrinter.cpp.

**Pull request merged**